pytorch_lightning/loops/utilities.py

```diff
@@ -139,3 +141,25 @@ def _block_parallel_sync_behavior(trainer: "pl.Trainer", block: bool = True) ->
             yield None
     else:
         yield None
+
+
+def _get_active_optimizers(
+    optimizers: List[Optimizer], frequencies: List[int], batch_idx: Optional[int] = None
+) -> List[Tuple[int, Optimizer]]:
+    """Returns the currently active optimizers. When multiple optimizers are used with different frequencies, only
+    one of the optimizers is active at a time.
+
+    Returns:
+        A list of tuples (opt_idx, optimizer) of currently active optimizers.
+    """
+    if not frequencies:
+        # call training_step once per optimizer
+        return list(enumerate(optimizers))
+
+    freq_cumsum = np.cumsum(frequencies)
+    optimizers_loop_length = freq_cumsum[-1]
+    current_place_in_loop = batch_idx % optimizers_loop_length
+
+    # find optimizer index by looking for the first {item > current_place} in the cumsum list
+    opt_idx = int(np.argmax(freq_cumsum > current_place_in_loop))
```

Comment:
since `freq_cumsum` will be sorted, this would be more efficient

```suggestion
    opt_idx = np.searchsorted(freq_cumsum, current_place_in_loop)
```

ref: https://stackoverflow.com/a/48360950

Reply:
this gave me a failing test: `tests/loops/optimization/test_optimizer_loop.py::test_optimizer_frequencies`
could potentially be added in a separate PR, I want to keep this a refactor